### PR TITLE
Add executable no-ai-attribution hook, closing the flag-level bypasses

### LINKED ISSUE #177: no-ai-attribution: attribution via flags (--trailer, --author, --template) and gh pr edit bypasses the message-text check
The `no-ai-attribution` hook spec keys on the *message text* of the git/gh subcommands it lists (`git commit -m`/`-F`, `gh pr create --body`/`--body-file`, the comment/review bodies). The cloud review on #176 surfaced four flag-level paths that introduce AI attribution without it ever appearing in that message text, so the message-text check does not see them:

- `git commit --trailer 'Co-Authored-By: Claude <...>'` — writes the trailer into the commit without a `-m`/`-F` attribution line to scan.
- `git commit --author='Claude <bot@example.com>'` — puts the model name in the commit's author field; the message itself is clean.
- `gh pr create --template <file>` (`-T`) — fills the PR body from a template file the command does not name as `--body-file`, so a template carrying a credit line is submitted unchecked.
- `gh pr edit --body`/`--body-file` — changes a PR body *after* creation; the spec only covers `gh pr create`.

These are the same bug class: attribution can enter the record through a flag or a later edit rather than through the inline/named-file message body the hook inspects.

Two ways to close it, for whoever implements the matcher (the hook currently ships as a markdown spec, no executable yet):
1. Parse the attribution-bearing flags too: scan `--trailer` and `--author` values on `git commit`, read `-T/--template` on `gh pr create`, and add `gh pr edit` to the covered subcommands.
2. Until then, the spec is scoped honestly to say the hook is a best-effort guard on the common command-line authoring path (message text of the listed subcommands), not a complete interception — done in the #176 doc update that references this issue.

Filed from the #176 cleanup so the gap is tracked rather than silently conceded.

### PR DESCRIPTION
**Pull request by jamditis**
## What this does

Adds the executable `no-ai-attribution` PreToolUse(Bash) hook that enforces the markdown spec of the same name and closes the four flag-level bypasses the spec could only describe: a commit `--author`/`--trailer` field, `gh pr create --template`, and `gh pr edit` bodies.

## Why

Authorship in a newsroom's git and PR record should name people, not tools. The prior guard was a spec with no enforcement. A spec can describe a bypass; only an executable can block it before the credit lands in history or on a merged PR.

## How it works

- Stdlib-only Python, one file for every OS. Reads the PreToolUse payload on stdin, exits `0` to allow or `2` with a reason on stderr to deny.
- Two detectors: a nuanced check for free prose (message, body, title, named file) that still allows naming a tool as the subject of the work, and a strict check for authorship fields where any tool token blocks.
- Fails open on any payload it cannot read (malformed JSON, unbalanced quotes, an unreadable file), so it never wedges a session, and says so on stderr rather than silently disabling.

## Bypasses closed beyond the four conceded ones

An adversarial review (local codex, gpt-5.5/low to convergence then gpt-5.6-sol/xhigh) surfaced shell-parsing gaps a spec never could, all now covered and tested: chained and glued operator separators (`a&&git commit`, `(cd x); git commit`), newline separators, env wrappers (identity vars, `-S`, `-C`, `--`, and attached forms), `git -c` identity config, ANSI-C `$'...'` quoted messages, and a relative message file resolved against the directory a `cd` or `git -C` would move into. Relative files are checked against both the tracked directory and the payload cwd, so a wrong guess about the run directory is a false positive at worst, never a missed byline.

## Known limitation

A watched command written inside a here-document body is checked as if it were a command (a safe-direction false positive). Documented in the spec's file-resolution section.

## How to verify

```
python -m pytest hooks/tests/test_no_ai_attribution.py -q
```

79 tests pin which command shapes block, which pass, and the fail-open cases.

Closes #177

### CODE REVIEW

**Comment by jamditis** [comment]
@codex review

File: hooks/no-ai-attribution.py [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+            seg_cwd = base
+            if spec is _GIT_COMMIT:
+                # -C composes onto the base, so a relative -F file resolves correctly.
+                seg_cwd = _git_effective_dir(core, start - 1, base)
+                # `GIT_AUTHOR_NAME=Claude git commit ...` sets authorship via the
+                # environment, the same as --author. Read those identity vars as fields.
+                for a in assigns:
+                    name, _, val = a.partition("=")
+                    if name in _GIT_IDENTITY_ENV and value_names_tool(val):
+                        return "tool or bot named in a git identity environment variable"
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:1283]: Fixed in 6db857f. Exported identity is now modeled as a last-wins map (dict), not an append list: a later `export` of the same var overwrites, and `unset` pops the key, so the commit is judged by the value the shell would actually use at commit time. An inline `VAR=... git commit` prefix overlays that map for the one command. Tests: `test_allow_exported_identity_overwritten_clean`, `test_allow_exported_identity_then_unset`, `test_allow_inline_prefix_overrides_exported_identity`, plus a blocking case for a single flagged export.

File: hooks/no-ai-attribution.py [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+    Only regular files are read, and only up to a byte cap: a FIFO or device would
+    block the read before the command runs, and an unbounded read of a huge file could
+    exhaust memory, either of which would wedge the session the hook must never wedge.
+    """
+    if name == "-":
+        return None
+    try:
+        p = Path(name)
+        if not p.is_absolute():
+            p = Path(cwd) / p
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:1048]: Fixed in 6db857f. The file loop now reads two interpretations of a `~`-leading name: the literal token and `os.path.expanduser(name)`. shlex(punctuation_chars=True) discards whether the `~` was quoted, so an unquoted `~/MSG` (shell-expanded) and a quoted `'~/MSG'` (literal cwd path) are indistinguishable after tokenization; reading both blocks if either holds a byline, and a wrong guess is a false positive, never a miss. `_read_named_file` stays pure-path so the quoted-literal target is still checked. Test: `test_block_tilde_quoted_real_target_in_cwd` plus the expanded-path case.

File: hooks/no-ai-attribution.py [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+                return True
+            if _AI_INLINE_RE.search(line) or _VERB_TOOL_RE.search(line):
+                return True
+    if _AI_BYLINE_RE.search(text):
+        return True
+    for m in _COAUTHOR_LINE_RE.finditer(text):
+        if value_names_tool(m.group("value")):
+            return True
+    if _VERB_TOOL_BYLINE_RE.search(text):
+        return True
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:217]: Intentionally left as-is, with the spec corrected to match. A bare tool name in free prose is subject matter, not a byline: a legitimate model-comparison line such as `- Claude` / `- GPT-4` in a commit body would false-block. A bare name is caught only when it lands in an authorship field (`value_names_tool` on author/committer/trailer values); the prose detector `contains_attribution` requires a verb plus with/by/using/via before the tool token. This tradeoff and the model-list example are documented in `no-ai-attribution.md`. Blocking bare names in prose would make the guard noisy enough to be disabled, which is the worse failure.

File: hooks/no-ai-attribution.py [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+                if _git_config_identity_hit(core, start - 1):
+                    return "tool or bot named in a git -c identity config value"
+            found = _collect(core[start:], spec)
+            for text in found["text"]:
+                # Check the raw token and, if it carries ANSI-C escapes, its decoded form:
+                # `-m $'...\nGenerated with Claude'` commits a real byline the raw token hides.
+                if contains_attribution(text):
+                    return "attribution in message/body text"
+                decoded = _decode_c_escapes(text)
+                if decoded != text and contains_attribution(decoded):
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in 6db857f via `_shell_message_variants(text)`, which yields the raw token, the `$`-stripped form, and the C-unescaped form. A whole-line ANSI-C message `$'...'` has its leading `$` marker peeled before the line-anchored byline check runs, not just its escapes decoded, so a byline on the first line is anchored correctly. Test covers the `$'...\n\n<byline>'` case.

File: hooks/no-ai-attribution.py [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+import sys
+from pathlib import Path
+
+HOOK_NAME = "no-ai-attribution"
+ROBOT_EMOJI = "\U0001F916"  # the sign-off emoji used as an AI byline
+
+# Tokens that name an AI tool, model, or vendor. Used with word boundaries.
+_TOOL_TOKENS = (
+    r"claude|chatgpt|gpt|codex|copilot|gemini|anthropic|openai|cursor|bard|llama|"
+    r"language model|large language model|llm"
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in 6db857f. The gpt token is now `gpt(?:-?[0-9][a-z0-9]*)?`, matching `GPT4` and `GPT-4o`, not only bare `gpt`. `\bgpt\b` failed because there is no word boundary between a letter and a digit, so `gpt4` slipped through. Test covers `GPT4` / `GPT-4o`.

File: hooks/no-ai-attribution.py [hunk 6]
```diff
@@ -0,0 +1,709 @@
...
+        return True
+    return False
+
+
+# --- command parsing -------------------------------------------------------
+
+# Tokens that separate one command from the next, or wrap a grouped command. A
+# command word right after any of these starts a fresh segment, so attribution in a
+# chained or grouped command (`a && git commit ...`, `(git commit ...)`) is still seen.
+_OPERATORS = {"&&", "||", ";", ";;", "|", "&", "|&", "(", ")", "{", "}"}
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:226]: Fixed in 6db857f. `_strip_leading_keywords` peels a leading shell-keyword set (`if elif then else while until do ! time`) from each segment before matching, so a one-line conditional like `if git commit -m '<byline>'; then ...` reaches the commit. Test covers the one-line conditional form.

File: hooks/no-ai-attribution.py [hunk 7]
```diff
@@ -0,0 +1,709 @@
...
+            # Resolve a relative file against both the tracked dir and the payload cwd, and
+            # block on either. cd tracking is a best effort (a subshell `(cd x)` does not
+            # persist, a failed `cd` does not move) so the shell's real cwd can differ from
+            # the tracked one; reading both makes a wrong guess a false positive, never a miss.
+            read_dirs = [seg_cwd] if seg_cwd == cwd else [seg_cwd, cwd]
+            for name in found["file"]:
+                for read_dir in read_dirs:
+                    content = _read_named_file(name, read_dir)
+                    if content is not None and contains_attribution(content):
+                        return f"attribution in the file '{name}'"
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Documented limitation, not fixable by a PreToolUse hook. The hook runs before the Bash command executes, so a message file the same invocation is about to create (`printf ... > MSG && git commit -F MSG`) does not exist yet to read: the same category as command substitution `$(...)`, whose value also does not exist pre-execution. Both are noted as limitations in `no-ai-attribution.md`.

File: hooks/no-ai-attribution.py [hunk 8]
```diff
@@ -0,0 +1,709 @@
...
+            letters = t[1:]
+            j = 0
+            consumed_next = False
+            while j < len(letters):
+                ch = letters[j]
+                kind = spec["short"].get(ch)
+                if kind:
+                    rest = letters[j + 1:]
+                    if rest:
+                        out[kind].append(rest[1:] if rest.startswith("=") else rest)
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in 6db857f. `_collect` no longer strips a leading `=` from an attached short-option value, matching git parse-options and gh/pflag, which keep it: `git commit -F=MSG` names the file `=MSG` and `-m=x` sets the message `=x`; only a long `--file=MSG` splits on `=`. So the hook now reads exactly the file/message git reads. Test covers `-F=MSG` and `-m=`.

File: hooks/no-ai-attribution.py [hunk 9]
```diff
@@ -0,0 +1,709 @@
...
+_BOT_MARKER_RE = re.compile(r"\bbot@|\[bot\]|noreply@anthropic|noreply@openai", re.I)
+
+
+# ANSI-C ($'...') escapes bash decodes before running the command. shlex keeps them as
+# literal backslash sequences, so `-m $'x\n\nGenerated with Claude'` reaches the record
+# with a real newline (a leading byline) while the hook sees a single literal line.
+_C_ESCAPES = {
+    "n": "\n", "t": "\t", "r": "\r", "f": "\f", "v": "\v",
+    "a": "\a", "b": "\b", "\\": "\\", "'": "'", '"': '"',
+}
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:139]: Fixed in 6db857f. `_VERB_TOOL_BYLINE_RE`'s leading character class is now `[ \t>*•:#_=-]*`, matching the AI-byline anchor set, so a verb-tool byline led by a bullet, colon, hash, underscore, or `=` marker is anchored and caught. Test covers the marker variants.

**Comment by jamditis** [comment]
@codex review

File: hooks/no-ai-attribution.py [hunk 10]
```diff
@@ -0,0 +1,1184 @@
...
+    """
+    assigns = []
+    unsets = []
+    chdir = None
+    i = 0
+    while i < len(seg) and _ENV_ASSIGN_RE.match(seg[i]):
+        assigns.append(seg[i])
+        i += 1
+    if i >= len(seg) or seg[i] != "env":
+        return assigns, seg[i:], chdir, unsets
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in a5f7963: `env` is now matched by basename, so `/usr/bin/env NAME=val git commit ...` is peeled the same as a bare `env`. Test: test_block_pathed_env_identity_prefix.

File: hooks/no-ai-attribution.py [hunk 11]
```diff
@@ -0,0 +1,1184 @@
...
+            core = _strip_command_wrappers(core)
+            if not core:
+                continue
+            # `bash -c '<script>'` hides a git/gh command in its script argument. Recurse
+            # into that script (depth-bounded) so an inline byline is caught the same as a
+            # top-level one; the runner itself writes nothing, so stop after recursing.
+            script = _shell_c_script(core)
+            if script is not None:
+                if depth < _MAX_SHELL_DEPTH:
+                    inner = find_attribution(script, eff_cwd, depth + 1)
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in a5f7963: the `bash -c` recursion now uses the `env -C`-adjusted cwd, so a relative file the inner command reads resolves in that dir. Test: test_block_env_C_before_shell_c_file.

File: hooks/no-ai-attribution.py [hunk 12]
```diff
@@ -0,0 +1,1184 @@
...
+_GH_EDIT = {  # gh pr edit / gh issue edit: title and body are both editable
+    "text": {"-b", "--body", "-t", "--title"},
+    "file": {"-F", "--body-file"},
+    "field": set(),
+    "short": {"b": "text", "t": "text", "F": "file"},
+}
+_GH_MERGE = {  # gh pr merge: --subject/--body set the merge/squash commit message
+    "text": {"-b", "--body", "-t", "--subject"},
+    "file": {"-F", "--body-file"},
+    "field": set(),
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in a5f7963: `-A`/`--author-email` is now a strict field on the merge spec, read like a git commit author. Tests: test_block_gh_pr_merge_author_email_long, _short.

File: hooks/no-ai-attribution.py [hunk 13]
```diff
@@ -0,0 +1,1184 @@
...
+            if fname and fname != "-" and _input_file_has_attribution(fname, read_dirs):
+                return "attribution in a gh api --input request body file"
+            continue
+        else:
+            i += 1
+            continue
+        if val is None:
+            continue
+        key, sep, v = val.partition("=")
+        if not sep or key.strip().lower() not in _GH_API_TEXT_KEYS:
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in a5f7963: `_gh_api_key_is_prose` normalizes bracketed keys, so a nested prose key such as `input[body]=` is scanned. Test: test_block_gh_api_nested_input_body.

File: hooks/no-ai-attribution.py [hunk 14]
```diff
@@ -0,0 +1,1184 @@
...
+            continue
+        # -F/--field reads a file when the value starts with `@` (@- is stdin, skipped).
+        if mode == "field" and v.startswith("@"):
+            fname = v[1:]
+            if fname and fname != "-" and _named_file_has_attribution(fname, read_dirs):
+                return "attribution in a gh api field file"
+            continue
+        if contains_attribution(v):
+            return "attribution in a gh api request field"
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in a5f7963: the gh api field path now runs the shell-message variants, decoding an ANSI-C ($'...') value the same as the commit-message path. Test: test_block_gh_api_ansi_c_field.

File: hooks/no-ai-attribution.py [hunk 15]
```diff
@@ -0,0 +1,1184 @@
...
+            continue
+
+        for seg in _segments(tokens):
+            seg = _strip_redirects(seg)
+            seg = _strip_leading_keywords(seg)
+            if not seg:
+                continue
+            assigns, core, env_chdir, unsets = _strip_env_prefix(seg)
+            if not core:
+                continue
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in a5f7963: a standalone reassignment of an already-exported identity var now updates the tracked value; a never-exported local assignment is still correctly ignored. Test: test_block_reassigned_exported_identity.

File: hooks/no-ai-attribution.py [hunk 16]
```diff
@@ -0,0 +1,1184 @@
...
+            if t.startswith("--unset="):
+                unsets.append(t.split("=", 1)[1])
+                j += 1
+                continue
+            if t in ("-u", "--unset"):
+                if j + 1 < len(rest):
+                    unsets.append(rest[j + 1])
+                j += 2
+                continue
+            j += 1  # any other env option takes no separate value token we track
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:799]: Fixed in a5f7963: `env -i`/`--ignore-environment` now starts the command with an empty identity map, so an identity an earlier export set is not counted; an inline assignment after -i is still read. Test: test_allow_env_i_clears_exported_identity.

File: hooks/no-ai-attribution.py [hunk 17]
```diff
@@ -0,0 +1,1184 @@
...
+        t = core[i]
+        if t == "--command" or t == "-c":
+            return core[i + 1] if i + 1 < len(core) else None
+        if t.startswith("--"):
+            i += 1
+            continue
+        if t.startswith("-") and len(t) > 1:
+            if "c" in t[1:]:  # -c in a cluster (e.g. -lc): the script is the next arg
+                return core[i + 1] if i + 1 < len(core) else None
+            i += 1
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:870]: Fixed in a5f7963: value-taking shell options (-o, -O, --rcfile, --init-file) are skipped before -c, so the script in `bash -o pipefail -c ...` is still found. Test: test_block_bash_o_pipefail_before_c.

File: hooks/no-ai-attribution.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in a5f7963: an explicit read-only method (--method GET / -X GET) is detected and its fields are left unscanned, since they are query parameters. Test: test_allow_gh_api_explicit_get_method.

**Comment by jamditis** [comment]
@codex review

**Comment by jamditis** [comment]
Consolidated the 10 open connector findings (shell-command parser edge cases: byline detection, command/segment boundaries, git-identity tracking) into #194 as a parser-hardening backlog. They are all P2/P3 hardening on a best-effort local guard, so they are tracked there with a verify-from-code + regression-test note rather than expanded into this PR's scope. Resolving these threads; they are filed, not dropped.

File: hooks/no-ai-attribution.py [hunk 18]
```diff
@@ -0,0 +1,1260 @@
...
+    fire.
+    """
+    if not text:
+        return False
+    # Robot-emoji byline: it leads a line (a sign-off marker position) or shares a line
+    # with an attribution cue. A robot emoji embedded in prose ("fix 🤖 rendering") is
+    # subject matter and does not fire.
+    for line in text.splitlines():
+        if ROBOT_EMOJI in line:
+            if line.lstrip().startswith(ROBOT_EMOJI):
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in 2b6ed70: the robot-emoji check now peels the same leading sign-off/list markers the text bylines ride (via `_BYLINE_LEAD_RE`), so `- 🤖` and `> 🤖` are caught. Tests: test_block_markdown_dash_robot_emoji_byline, test_block_quote_marker_robot_emoji_byline.

File: hooks/no-ai-attribution.py [hunk 19]
```diff
@@ -0,0 +1,1260 @@
...
+            i += 2
+        elif t.startswith("-c") and not t.startswith("--") and len(t) > 2:
+            cfg = t[2:]  # attached form: -cuser.name=Claude
+            i += 1
+        else:
+            i += 1
+            continue
+        name, sep, val = cfg.partition("=")
+        if sep and name.strip().lower() in _GIT_IDENTITY_CONFIG and value_names_tool(val):
+            return True
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py]: Fixed in 2b6ed70: `_git_config_identity_hit` now collects the last value per identity key (git applies last-wins), so a clean override of an earlier tool name passes and the tool name winning last still blocks. Tests: test_allow_git_c_identity_last_wins_clean, test_block_git_c_identity_last_wins_tool.

File: hooks/no-ai-attribution.py [hunk 20]
```diff
@@ -0,0 +1,1260 @@
...
+# classification. Peel them so the command is still seen. A keyword appearing mid-segment
+# is an ordinary argument (a word in a commit message) and is left untouched.
+_SHELL_KEYWORDS = {"if", "elif", "then", "else", "while", "until", "do", "!", "time"}
+
+
+def _strip_leading_keywords(seg):
+    """Drop any leading shell control keywords from a segment, returning the remainder."""
+    i = 0
+    while i < len(seg) and seg[i] in _SHELL_KEYWORDS:
+        i += 1
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:537]: Fixed in 2b6ed70: `_strip_leading_keywords` now skips the optional `-p` after `time`, so `time -p git commit ...` exposes the command word. Test: test_block_time_dash_p_wrapper_commit (with test_allow_time_dash_p_clean_commit as a guard).

File: hooks/no-ai-attribution.py [hunk 21]
```diff
@@ -0,0 +1,1260 @@
...
+                # `export GIT_AUTHOR_NAME=Claude; git commit` sets the identity in this
+                # shell for the later commit, the same authorship surface as an inline
+                # `GIT_AUTHOR_NAME=Claude git commit` prefix. Track the latest value per
+                # name (a re-export overwrites, an `unset` below clears) so a later clean
+                # value is not shadowed by an earlier flagged one. (A bare `export NAME`
+                # with no value, or an export inside a subshell, is best-effort at worst.)
+                for t in core[1:]:
+                    if _ENV_ASSIGN_RE.match(t):
+                        name, _, val = t.partition("=")
+                        exported_identity[name] = val
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:1244]: Fixed in 2b6ed70: a bare `export NAME` now marks the name for export (`setdefault`), so a later standalone `NAME=value` on its own line updates the tracked identity. Test: test_block_export_bare_then_assign_identity (with test_allow_export_bare_then_clean_assign as a guard).

File: hooks/no-ai-attribution.py [hunk 22]
```diff
@@ -0,0 +1,1260 @@
...
+def _tokenize(command):
+    """Tokenize a shell command, keeping operators separate even without spaces.
+
+    Uses shlex with punctuation_chars so `a&&b` splits into `a`, `&&`, `b` (which
+    plain shlex.split does not), while operators and `#` inside quotes stay part of
+    their token. Raises ValueError on unbalanced quoting, which the caller handles.
+    """
+    lex = shlex.shlex(command, posix=True, punctuation_chars=True)
+    lex.whitespace_split = True
+    lex.commenters = ""  # a commit message is not a script; do not treat # as a comment
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:255]: Fixed in 2b6ed70: a new quote-aware `_strip_comment` drops an unquoted, word-start `#` comment before parsing, so a clean command with an attributed example in a trailing comment is no longer blocked; a `#` inside quotes or glued mid-word stays literal. Tests: test_allow_trailing_comment_with_attributed_example, test_block_hash_inside_quoted_message_still_scanned, test_allow_hash_midword_in_message.

File: hooks/no-ai-attribution.py [hunk 23]
```diff
@@ -0,0 +1,1260 @@
...
+            # cd tracking is best-effort (a subshell `(cd x)` does not persist, a failed cd
+            # does not move), so resolve a relative file against both the tracked dir and
+            # the payload cwd; a wrong guess is a false positive, never a miss.
+            read_dirs = [seg_cwd] if seg_cwd == cwd else [seg_cwd, cwd]
+            if spec is _GH_API:
+                reason = _gh_api_hits(core, read_dirs)
+                if reason is not None:
+                    return reason
+                continue
+            found = _collect(core[start:], spec)
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:1297]: Fixed in 2b6ed70: `_git_commit_is_dry_run` detects `--dry-run` (and its unambiguous abbreviation) and skips the segment, since a dry-run creates no commit. A chained real commit after it still blocks. Tests: test_allow_dry_run_attributed_message, test_allow_dry_run_abbrev_message, test_block_no_verify_is_not_dry_run, test_block_real_commit_chained_after_dry_run.

File: hooks/no-ai-attribution.py [hunk 24]
```diff
@@ -0,0 +1,1260 @@
...
+
+
+def _tokenize(command):
+    """Tokenize a shell command, keeping operators separate even without spaces.
+
+    Uses shlex with punctuation_chars so `a&&b` splits into `a`, `&&`, `b` (which
+    plain shlex.split does not), while operators and `#` inside quotes stay part of
+    their token. Raises ValueError on unbalanced quoting, which the caller handles.
+    """
+    lex = shlex.shlex(command, posix=True, punctuation_chars=True)
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:253]: Documented as a known limitation in 2b6ed70 (spec "What it does not catch"): an escaped apostrophe inside an ANSI-C `$'...'` quote makes the tokenizer reject the token, and the raw-text fallback leaves the literal `\n` undecoded. The common ANSI-C forms without an escaped inner quote are decoded and checked; this one spelling needs a form no ordinary workflow produces, and the layers around the hook cover it.

File: hooks/no-ai-attribution.py [hunk 25]
```diff
@@ -0,0 +1,1260 @@
...
+    """Return the effective directory after a `cd` segment.
+
+    A relative message file resolves against the shell's current directory, so a
+    preceding `cd repo` must move the base a `git commit -F MSG` reads MSG from. `cd -`
+    (previous dir) is unknowable and leaves the base unchanged; a bare `cd` goes home.
+    """
+    target = None
+    for t in core[1:]:
+        if t == "--" or (t.startswith("-") and t != "-"):
+            continue
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:1143]: Documented as a known limitation in 2b6ed70: a directory whose name starts with `-` (`cd -- -repo`) is read by the cd tracker as an option and not followed, so a relative message file there resolves from the wrong directory. A leading-dash directory name is not a form an ordinary workflow produces; the surrounding layers cover it.

File: hooks/no-ai-attribution.py [hunk 26]
```diff
@@ -0,0 +1,1260 @@
...
+                if depth < _MAX_SHELL_DEPTH:
+                    # `env -C dir bash -c '...'` runs the script in dir, so a relative file
+                    # the inner command reads (git commit -F MSG) resolves there. Recurse
+                    # with the env -C-adjusted cwd, not the bare tracked cwd.
+                    rec_cwd = eff_cwd if env_chdir is None else _resolve_dir(eff_cwd, env_chdir)
+                    inner = find_attribution(script, rec_cwd, depth + 1)
+                    if inner is not None:
+                        return inner
+                continue
+            if core[0] == "cd":
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:1231]: Documented as a known limitation in 2b6ed70: only `cd` updates the tracked directory, so `pushd repo && git commit -F MSG` is not followed (the file still blocks if a same-named file in the payload cwd carries a byline). Tracking pushd/popd is out of scope for the accidental-attribution path the hook guards.

File: hooks/no-ai-attribution.py [hunk 27]
```diff
@@ -0,0 +1,1260 @@
...
+def find_attribution(command, cwd, depth=0):
+    """Return a short reason string if the command would write AI attribution, else None.
+
+    Walks the command's logical lines (so a newline-separated command is seen), then the
+    segments of each, tracking the effective directory across `cd` so relative message
+    files resolve where the command would read them. `depth` bounds recursion into a shell
+    runner's `-c` script (bash -c '<script>'), so a pathological nest cannot run away.
+    """
+    eff_cwd = cwd
+    exported_identity = {}  # exported identity var -> latest value (shell last-wins)
```
[the remark this answers is not in the document]
Comment (jamditis) [reply hooks/no-ai-attribution.py:1183]: Documented as a known limitation in 2b6ed70: the identity check reads the assignments in the command string, not the process environment, so a `GIT_COMMITTER_NAME`/`GIT_AUTHOR_NAME` exported before the hook runs is not seen. Reading ambient os.environ is a separate surface with its own false-positive risk; this inherited-state path is covered by the layers around the hook.
